Update to 3.2.4

wshanks opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->
This is a continuation of #37 (since I can't push to it). The project switched to meson-python from setuptools, so I switched out the host dependency. Also, it renamed the git submodule from scs to scs_source, so I also changed the source directory in the recipe and updated the commit to match upstream. There was a post release that updated how the blas libraries are detected with meson and conda-build also needs this update, so the .post1 source is used.

The ppc64le pypy build has some issue. I noticed that the scipy feedstock disabled this build because this combination is not well maintained, so I also disabled it instead of trying to debug the build issue.

My main motivation here is to get a Python 3.12 build. 3.2.4 added support for Python 3.12.

Closes #37, #36

<!--
Please add any other relevant info below:
-->

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -1,35 +1,38 @@
 source:
-  - url: https://github.com/bodono/scs-python/archive/refs/tags/{{ version }}.tar.gz
-    sha256: 7592805810fb6da09955dc45322b2d376bca993928c972f472f1a5352821ddcd
+  - url: https://github.com/bodono/scs-python/archive/refs/tags/{{ version }}.post1.tar.gz
+    sha256: bb509507d857906eb95d17d5824138e48e7c75d1a692d0aa7256d899c7e703c5
```
Please add the `.post1` to the version rather than to the URL

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -1,35 +1,38 @@
-    - export BLAS_LAPACK_LIB_PATHS="$PREFIX/lib"    # [unix]
-    - export BLAS_LAPACK_LIBS="lapack:cblas:blas"   # [unix]
-    - {{ PYTHON }} -m pip install . -vv
+    # - export PKG_CONFIG_PATH=$PREFIX/lib/pkgconfig  # [ppc64le]
+    # - echo "python = '${PREFIX}/bin/python'" >> ${CONDA_PREFIX}/meson_cross_file.txt  # [ppc64le]
+    - MESON_ARGS_REDUCED="$(echo $MESON_ARGS | sed 's/--buildtype release //g')"  # [build_platform != target_platform]
+    - {{ PYTHON }} -m pip install . -vv -Csetup-args=${MESON_ARGS_REDUCED// / -Csetup-args=}  # [build_platform != target_platform]
+    - {{ PYTHON }} -m pip install . -vv  # [build_platform == target_platform]
```
Please remove debug output, align selectors, and perhaps add a comment about why we need to strip that option (if you found that being used somewhere else in conda-forge already, just copy the comment from there too; if you came up with it yourself, the comment shouldn't be hard ;-)).

wshanks replied:
Thanks, I will clean this up. It worked locally for me on Linux. I wasn't expecting the trouble with cross-compiling when I opened the PR. I will mark it as a draft for now.

h-vetinari reviewed (commented):
Thanks for working on this!

You'll very likely have to pass some blas/lapack options to meson as well (you can use the scipy feedstock for reference/inspiration on this).

```
  Library blas found: NO
  Run-time dependency blas found: NO (tried pkgconfig and cmake)
  Library lapack found: NO
  Run-time dependency lapack found: NO (tried pkgconfig and cmake)
  Library cblas found: NO
  Run-time dependency cblas found: NO (tried pkgconfig and cmake)

../meson.build:49:4: ERROR: Problem encountered: OpenBLAS or Netlib BLAS/CBLAS is required on all platforms except Windows, and was not found.
```

wshanks commented:
@h-vetinari I got everything green now. Part of my issue earlier was that I didn't realize that Linux aarch64 was cross-compiled so my `build_platform != target_platform` was covering more cases than I realized (I thought it was only macOS ARM64), so I kept adjust other things back and forth fixing osx and breaking linux and vice versa.

What I have ended up with here is what empirically I found to work. The `pip install .` worked (CI green here; I didn't test locally other than Linux x86_64) for everything except osx_arm64. For osx_arm64, I copied what pandas did to modify the meson args. That modification didn't work for Linux aarch64, so I split out just osx_arm64 (for pandas all cross-compiled builds use the meson args adjustment). I didn't debug exactly why the meson args adjustment is needed for osx_arm64 but breaks linux_aarch64. I dropped the ppc64le pypy build because it failed without an obvious error message and I saw you had also dropped it from scipy because that combination is not well maintained.

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -1,35 +1,46 @@
+    # instead of unix/win because Linux aarch64 does not work with the Meson
+    # args modification (fails to find blas/lapack)
+    #
+    # See https://github.com/conda-forge/pandas-feedstock/blob/ea054ffba9cda5f10436e1c7250e62d7ee1aacd1/recipe/meta.yaml#L20-L22
+    - MESON_ARGS_REDUCED="$(echo $MESON_ARGS | sed 's/--buildtype release //g')"              # [not aarch64 and build_platform != target_platform]
+    - {{ PYTHON }} -m pip install . -vv -Csetup-args=${MESON_ARGS_REDUCED// / -Csetup-args=}  # [not aarch64 and build_platform != target_platform]
+    - {{ PYTHON }} -m pip install . -vv                                                       # [aarch64 or build_platform == target_platform]
```
What I meant is that we should explicitly tell meson to look for the right lapack:
```
    -Csetup-args=-Dblas=blas \
    -Csetup-args=-Dlapack=lapack \
```
cf. scipy's [`build.sh`](https://github.com/conda-forge/scipy-feedstock/blob/main/recipe/build.sh). I don't want to special case aarch64 here without trying to fix it properly first.

wshanks replied:
I believe that osx_arm64 is the special case. From what I saw, ppc64le worked with the MESON_ARGS_REDUCED commands as well as the simple `pip install .`. So we could rewrite the selectors as `osx and arm64` and `not (osx and arm64)`. Would that be better?

There are a few things I am unclear on currently. With the current configuration, I see for all builds that I checked:

```
  Library blas found: NO
  Run-time dependency blas found: YES 3.9.0
  Library lapack found: NO
  Run-time dependency lapack found: YES 3.9.0
  Library cblas found: NO
  Run-time dependency cblas found: YES 3.9.0
```

Are the NO answers here an issue? It does find the run-time dependencies and build and pass the tests. I tried adding the blas and lapack options [here](https://github.com/conda-forge/scs-feedstock/commit/60f4907ed33bcbbda44e81e7d304ba7da13f036c) but got

```
../meson.build:1:0: ERROR: Unknown options: "blas, lapack"
```

Maybe I did not do it right? I interpreted that to mean those were meson config options that scs-python had not set up.

I don't fully understand what `MESON_ARGS_REDUCED` is doing. From comment in the pandas feedstock (really more [this comment](https://github.com/conda-forge/scipy-feedstock/blob/9401aa4ee5d9a22c2c9d219e5fb6dbd580f30207/recipe/build.sh#L12-L14) in the scipy feedstock), I took it that the buildtype flag was already inserted into `MESON_ARGS` by meson-python and having it included twice was problematic so it needs to be removed. However, if I don't use the `not aarch64` selector (like [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=862809&view=logs&jobId=e9afaa34-1a0f-534e-78fc-fae528ccd915&j=e9afaa34-1a0f-534e-78fc-fae528ccd915&t=869ab379-d63c-50ae-655f-69018fb07995)), then I see for aarch64

```
  Library blas found: NO
  Run-time dependency blas found: NO (tried pkgconfig and cmake)
  Library lapack found: NO
  Run-time dependency lapack found: NO (tried pkgconfig and cmake)
  Library cblas found: NO
  Run-time dependency cblas found: NO (tried pkgconfig and cmake)
```

and the aarch64 build fails completely. I don't understand what failing to find blas/lapack has to do with filtering `buildtype` out of the meson args.

Also, if I don't use the `MESON_ARGS_REDUCED` method to filter out `buildtype` (like [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=863496&view=logs&jobId=1b8be447-c2bd-5772-b66a-a1146441bf88&j=1b8be447-c2bd-5772-b66a-a1146441bf88&t=0da7bde9-21e8-5d19-f418-738876121ca7)), the osx_arm64 build fails with:

  ../meson.build:1:0: ERROR: Could not invoke sanity test executable: [Errno 86] Bad CPU type in executable: '/Users/runner/miniforge3/conda-bld/scs_1706040399741/work/.mesonpy-n38da1p9/meson-private/sanitycheckc.exe'.
```

That seems like the cross-compilation is not configured properly and a build meson executable is getting used instead of the host one, but I don't see what that has to do with filtering out the `buildtype` (unless cross-compilation is an option for `buildtype`, but it seems like it is for release/debug).

h-vetinari replied:
> That seems like the cross-compilation is not configured properly and a build meson executable is getting used instead of the host one, but I don't see what that has to do with filtering out the `buildtype` (unless cross-compilation is an option for `buildtype`, but it seems like it is for release/debug).

Most likely you need something like [this](https://github.com/conda-forge/scipy-feedstock/blob/main/recipe/build.sh#L6-L10).

> Are the NO answers here an issue? It does find the run-time dependencies and build and pass the tests. I tried adding the blas and lapack options [here](https://github.com/conda-forge/scs-feedstock/commit/60f4907ed33bcbbda44e81e7d304ba7da13f036c) but got
> 
> ```
> ../meson.build:1:0: ERROR: Unknown options: "blas, lapack"
> ```

Not 100% sure if this needs handling in scs's meson files. CC the people involed in https://github.com/bodono/scs-python/pull/70: @enzbus @rgommers @bodono

> However, if I don't use the `not aarch64` selector (like [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=862809&view=logs&jobId=e9afaa34-1a0f-534e-78fc-fae528ccd915&j=e9afaa34-1a0f-534e-78fc-fae528ccd915&t=869ab379-d63c-50ae-655f-69018fb07995)), then I see for aarch64 [...]

I haven't had time to dig yet, but my point is: the only distinction in the builds here should be between native compilation and cross-compilation, not between linux-aarch64 & osx-arm64.

enzbus replied:
The NO answer by scs-python meson file are not an issue, if it compiles and passes tests it's all good. (We have a few library searches to catch various configurations.)

enzbus replied:
And for osx, it should link to the system libraries (Accelerate) automatically, both amd64 and arm64...

enzbus replied:
PS @wshanks , it looks like we didn't implement in the `meson.option` an arbitrary search path for (open)blas, feel free to open a PR in scs-python if you need that. Here https://github.com/bodono/scs-python/blob/dd17e2e5282ebe85f2df8a7c6b25cfdeb894970d/meson.build#L42 we hard-code a path that's used in github CI to build windows wheels. Maybe we can add another block with a similar syntax for the path you need. I'm not an expert in meson, from a quick search it appears you can export search paths directly to the compiler if you need -> https://stackoverflow.com/questions/61058722/how-can-i-specify-library-path-when-using-meson . Not sure if that helps, let me know!

wshanks replied:
@enzbus I don't think passing a specific path should be necessary. Looking at scipy, I see that it just uses setup options like `-Dblas=blas` here:

https://github.com/conda-forge/scipy-feedstock/blob/9401aa4ee5d9a22c2c9d219e5fb6dbd580f30207/recipe/build.sh#L17-L20

but scs should already be trying `blas=blas` and `lapack=lapack` here:

https://github.com/bodono/scs-python/blob/dd17e2e5282ebe85f2df8a7c6b25cfdeb894970d/meson.build#L10-L25

For reference the `blas` and `lapack` options get defined for scipy:

https://github.com/scipy/scipy/blob/e8b043df1d586d9df9fed96add1a536859f1cb8c/meson_options.txt#L1-L4

and then used around here:

https://github.com/scipy/scipy/blob/e8b043df1d586d9df9fed96add1a536859f1cb8c/scipy/meson.build#L165

The usage is ultimately just calling `dependency('blas')` which is what scs does as well, so I think the look up should be the same as for scipy.

The only significant differences that I see between this feedstock and the scipy feedstock currently are that the scipy feedstock includes pkgconfig (the blas lookup failure mentions that it tried using pkgconfig) and the scipy feedstock uses `python -m build` instead of `pip install`. I think this is because pip used to take only the last `-C` option rather than concatenating them. There is a warning about this [here](https://meson-python.readthedocs.io/en/latest/how-to-guides/config-settings.html) but the linked issue in pip was closed as resolved last year. I already tried adding pkgconfig before and did not see a change. Maybe there was some other issue at the time. I will try adding pkgconfig again and using `python -m build` to see if they help.

I am also confused about why the build works for aarch64 without taking `--buildtype release` out of the MESON_ARGS. For the PyPy ppc issue commented on elsewhere in the PR, I tried printing out `importlib.machinery.EXTENSION_SUFFIXES` and for aarch64 it printed out an x86_64 suffix which made me confused about how the cross-compile is set up. For ppc, the suffix was a ppc suffix, not x86.

rgommers replied:
Installing `pkgconfig` should make `dependency('blas')` work in conda-forge.

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -1,35 +1,46 @@
+    - {{ PYTHON }} -m pip install . -vv                                                       # [aarch64 or build_platform == target_platform]
+  # Following scipy feedstock (see https://github.com/conda-forge/scipy-feedstock/blob/4fde29215004ce771f5407153e4df438b76afa57/recipe/meta.yaml#L43-L45):
+  #
+  # pypy on aarch/ppc is un(der)maintained and currently broken, see
+  # https://github.com/conda-forge/pypy3.6-feedstock/issues/111
+  skip: true  # [ppc64le and python_impl == "pypy"]
```
What was the issue with pypy on ppc? Also, the skip only affects ppc but the comment talks about aarch/ppc

wshanks replied:
Yes, I could remove the aarch part. I took the three lines you highlighted unedited from the link I gave just above, so it was a direct quotation. I left the aarch64 pypy build since it didn't fail.

The ppc failure looks like [this](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=863592&view=logs&j=b7a895b1-8909-5023-4a22-40a4698fe576&t=821caef9-b8fa-5ab7-47a2-7e7cdae7737e&l=247):

```
    File "/home/conda/feedstock_root/build_artifacts/scs_1706047896693/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/pypy3.9/site-packages/mesonpy/__init__.py", line 78, in <module>
      assert all(re.match(_EXTENSION_SUFFIX_REGEX, f'foo{x}') for x in importlib.machinery.EXTENSION_SUFFIXES)
```

That is failing [this assertion](https://github.com/mesonbuild/meson-python/blob/9d0fa1336fe18ffcd258b31c327ff3c5a020532e/mesonpy/__init__.py#L78-L79) in meson-python. It seems that ppc/pypy includes an extension module file suffix in `importlib.machinery.EXTENSION_SUFFIXES` that meson-python does not expect? I am not sure if that is ppc/pypy's fault or meson-python's. It may be worth reporting to meson-python.

wshanks replied:
I opened an issue with meson-python about this here: https://github.com/mesonbuild/meson-python/issues/563.

h-vetinari reviewed (commented):
Still interested in fixing things more comprehensively than just special-casing aarch.

wshanks commented:
@h-vetinari I think this looks reasonable now (you might quibble with the formatting but I think the basic commands are reasonable). I am not really sure why I didn't get this working earlier. I did try all the changes here previously and I thought I tried them together but I must have had something off and then gone down several wrong directions. Some of the old Azure build logs are already missing, so I can't easily check what led me off track now.

The changes I needed were:

* Update the source url plus the scs submodule url and adjust the directory name for the scs submodule
* Switch setuptools to meson-python
* Include pkg-config so aarch64 can find blas libraries
* Include the meson_cross_file hack
* Include the MESON_ARGS buildtype filtering
* Skip the ppc pypy build

One thing that I think might have tripped me up is that if you don't do some of the steps above the cross builds end up building the native package instead and looking like they succeed on some platforms, so I was switching back and forth between osx_arm64 and aarch64 looking green but maybe they were both not really right.

h-vetinari reviewed (approved):
Cool, thanks for sticking with this, and the inputs from @enzbus @rgommers!

h-vetinari commented:
If you want to join as a maintainer here, please open a PR (or issue with a bot command) to add yourself!

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,47 @@
-    - {{ PYTHON }} -m pip install . -vv
+    # Copy cross-compile meson file hack from scipy feedstock:
+    # See https://github.com/conda-forge/scipy-feedstock/blob/9401aa4ee5d9a22c2c9d219e5fb6dbd580f30207/recipe/build.sh#L10
+    - echo "python = '${PREFIX}/bin/python'" >> ${CONDA_PREFIX}/meson_cross_file.txt
```
This isn't a huge issue and I didn't want to hold things up further, but this should really have at least a `# [unix]` selector, or even better: `# [build_platform != target_platform]`. IIRC the cross-file is only created in that case, and we shouldn't write garbage in there that might end up confusing meson (the reason why it works is that we're not specifying `--cross-file=...` in the native case, so the incomplete config there doesn't get picked up).
